tuplex/codegen/src/IteratorContextProxy.cc

```diff
@@ -0,0 +1,494 @@
+
+            BasicBlock *currBB = builder.GetInsertBlock();
+            BasicBlock *notExhaustedBB = BasicBlock::Create(_env->getContext(), "notExhaustedBB", currBB->getParent());
+            BasicBlock *defaultArgBB = BasicBlock::Create(_env->getContext(), "defaultArgBB", currBB->getParent());
+            BasicBlock *endBB = BasicBlock::Create(_env->getContext(), "endBB", currBB->getParent());
+
+            auto exhausted = updateIteratorIndex(builder, iterator, iteratorInfo);
+            // if a default value is provided, use phi nodes to choose from value based on index (iterator not exhausted) or default value (iterator exhausted)
+            // else check for exception and return value based on index if iterator not exhausted
+            if(defaultArg.val) {
+                builder.CreateCondBr(exhausted, defaultArgBB, notExhaustedBB);
+            } else {
+                lfb.addException(builder, ExceptionCode::STOPITERATION, exhausted);
+                builder.CreateBr(notExhaustedBB);
+            }
+
+            builder.SetInsertPoint(notExhaustedBB);
+            auto nextVal = getIteratorNextElement(builder, yieldType, iterator, iteratorInfo);
+            llvm::Value *retValNotExhausted = nextVal.val, *retSizeNotExhausted = nextVal.size;
+            builder.CreateBr(endBB);
+
+            builder.SetInsertPoint(defaultArgBB);
+            builder.CreateBr(endBB);
+
+            builder.SetInsertPoint(endBB);
+            lfb.setLastBlock(endBB);
+            if(defaultArg.val) {
+                auto retVal = builder.CreatePHI(_env->pythonToLLVMType(yieldType), 2);
+                auto retSize = builder.CreatePHI(_env->i64Type(), 2);
+                retVal->addIncoming(retValNotExhausted, notExhaustedBB);
+                retSize->addIncoming(retSizeNotExhausted, notExhaustedBB);
+                retVal->addIncoming(defaultArg.val, defaultArgBB);
+                retSize->addIncoming(defaultArg.size, defaultArgBB);
+                return SerializableValue(retVal, retSize);
+            } else {
+                return SerializableValue(retValNotExhausted, retSizeNotExhausted);
+            }
+        }
+
+        llvm::Value *IteratorContextProxy::updateIteratorIndex(llvm::IRBuilder<> &builder,
+                                                               llvm::Value *iterator,
+                                                               IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            llvm::Type *iteratorContextType = iterator->getType()->getPointerElementType();
+            std::string funcName;
+            auto iteratorName = iteratorInfo->iteratorName;
+
+            if(iteratorName == "zip") {
+                return updateZipIndex(builder, iterator, iteratorInfo);
+            }
+
+            if(iteratorName == "enumerate") {
+                return updateEnumerateIndex(builder, iterator, iteratorInfo);
+            }
+
+            assert(iteratorName == "iter");
```

Comment:
I think this assert should get converted to an if, and whenever an unknown iterator name is used this function should throw an error and use fallback mode.

Reply:
Added error message.